packages/flutter/lib/src/widgets/sliver.dart

```diff
@@ -966,35 +1006,89 @@ class SliverMultiBoxAdaptorElement extends RenderObjectElement implements Render
     super.performRebuild();
-    _currentBeforeChild = null;
     assert(_currentlyUpdatingChildIndex == null);
     try {
+      final Set<int> affectedIndexes = {};
+      final Map<int, Element> newKeepAliveChildren = <int, Element>{};
+
+      int getNewIndex(Widget w){
+        if (w.key != null){
+          return widget.delegate.getIndexByKey(w.key);
+        } else {
+          return -1;
+        }
+      }
+
       void processElement(int index) {
         _currentlyUpdatingChildIndex = index;
         final Element newChild = updateChild(_childElements[index], _build(index), index);
         if (newChild != null) {
           _childElements[index] = newChild;
-          final SliverMultiBoxAdaptorParentData parentData = newChild.renderObject.parentData;
-          if (!parentData.keptAlive)
-            _currentBeforeChild = newChild.renderObject;
         } else {
           _childElements.remove(index);
         }
       }
-      // processElement may modify the Map - need to do a .toList() here.
-      _childElements.keys.toList().forEach(processElement);
+
+      void removeChildIfNotRemapped(int index){
+        final Element child = _childElements[index];
+        final SliverMultiBoxAdaptorParentData originalParentData = child?.renderObject?.parentData;
+        if (originalParentData != null){
+          if (!originalParentData.keepAlive || getNewIndex(
+              child.widget) == -1) {
+            _childElements[index] = updateChild(_childElements[index], null, index);
+          }
+        }
+      }
+
+      void remappingKeepAliveChildren(int index) {
+        _currentlyUpdatingChildIndex = index;
+        if (_childElements[index] != null) {
+          removeChildIfNotRemapped(index);
+        }
+        final Element newChild = updateChild(newKeepAliveChildren[index], _build(index), index);
+        if (newChild != null) {
+          _childElements[index] = newChild;
+        } else {
+          _childElements.remove(index);
+        }
+      }
+
+      void remappingKeepAliveChildIndexes(int index){
+        final SliverMultiBoxAdaptorParentData originalParentData = _childElements[index]?.renderObject?.parentData;
+        if (originalParentData != null && originalParentData.keepAlive) {
+          final int newIndex = getNewIndex(_childElements[index].widget);
+          if(newIndex!=-1 && newIndex != index){
+            newKeepAliveChildren[newIndex] = _childElements[index];
+            _childElements.remove(index);
+            affectedIndexes.add(newIndex);
+          }
+        }
+      }
+      // children might change during remapping
+      List<int> remainingIndexes = _childElements.keys.toList();
+      _childElements.keys.toList().forEach(remappingKeepAliveChildIndexes);
+      affectedIndexes.toList().forEach(remappingKeepAliveChildren);
+
+      remainingIndexes = remainingIndexes.where((int i) => !affectedIndexes.contains(i))
+          .toList();
```

Comment:
You don't need the toList call here since you just iterate over everything right below.

Reply:
refactored